service/rpccommon/server.go

```diff
@@ -0,0 +1,343 @@
+	if !logEnabled {
+		log.SetOutput(ioutil.Discard)
+	}
+	if config.ApiVersion < 2 {
+		log.Printf("Using API v1")
+	}
+	return &ServerImpl{
+		config:   config,
+		listener: config.Listener,
+		stopChan: make(chan struct{}),
+	}
+}
+
+func (s *ServerImpl) Stop(kill bool) error {
+	if s.config.AcceptMulti {
+		close(s.stopChan)
+		s.listener.Close()
+	}
+	return s.debugger.Detach(kill)
+}
+
+func (s *ServerImpl) Restart() error {
+	if s.config.AttachPid != 0 {
+		return errors.New("cannot restart process Delve did not create")
+	}
+	return s.s2.Restart(rpc2.RestartIn{}, nil)
+}
+
+// Run starts a debugger and exposes it with an HTTP server. The debugger
+// itself can be stopped with the `detach` API. Run blocks until the HTTP
+// server stops.
+func (s *ServerImpl) Run() error {
+	var err error
+
+	if s.config.ApiVersion < 2 {
+		s.config.ApiVersion = 1
+	}
+	if s.config.ApiVersion > 2 {
+		return fmt.Errorf("unknown API version")
+	}
+
+	// Create and start the debugger
+	if s.debugger, err = debugger.New(&debugger.Config{
+		ProcessArgs: s.config.ProcessArgs,
+		AttachPid:   s.config.AttachPid,
+	}); err != nil {
+		return err
+	}
+
+	s.s1 = rpc1.NewServer(s.config, s.debugger)
+	s.s2 = rpc2.NewServer(s.config, s.debugger)
+
+	rpcServer := &RPCServer{s}
+
+	s.methodMaps = make([]map[string]*methodType, 2)
+
+	s.methodMaps[0] = map[string]*methodType{}
+	s.methodMaps[1] = map[string]*methodType{}
+	suitableMethods(s.s1, s.methodMaps[0])
+	suitableMethods(rpcServer, s.methodMaps[0])
+	suitableMethods(s.s2, s.methodMaps[1])
+	suitableMethods(rpcServer, s.methodMaps[1])
+
+	go func() {
+		defer s.listener.Close()
+		for {
+			c, err := s.listener.Accept()
+			if err != nil {
+				select {
+				case <-s.stopChan:
+					// We were supposed to exit, do nothing and return
+					return
+				default:
+					panic(err)
+				}
+			}
+			go s.serveCodec(c)
+			if !s.config.AcceptMulti {
+				break
+			}
+		}
+	}()
+	return nil
+}
+
+// Precompute the reflect type for error.  Can't use error directly
+// because Typeof takes an empty interface value.  This is annoying.
+var typeOfError = reflect.TypeOf((*error)(nil)).Elem()
+
+// Is this an exported - upper case - name?
+func isExported(name string) bool {
+	rune, _ := utf8.DecodeRuneInString(name)
+	return unicode.IsUpper(rune)
+}
+
+// Is this type exported or a builtin?
+func isExportedOrBuiltinType(t reflect.Type) bool {
+	for t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+	// PkgPath will be non-empty even for an exported type,
+	// so we need to check the type name as well.
+	return isExported(t.Name()) || t.PkgPath() == ""
+}
+
+// Fills methods map with the methods of receiver that should be made
+// available through the RPC interface.
+// These are all the public methods of rcvr that have one of those
+// two signatures:
+// func (rcvr ReceiverType) Method(in InputType, out *ReplyType) error
+// func (rcvr ReceiverType) Method(in InputType, cb service.RPCCallback
```

Comment:
Need to fix example func signature on this line. (Missing ending paran and return type).